libr/core/rvc.c

```diff
@@ -66,491 +32,306 @@ static inline char *sha256_file(const char *fname) {
 	return find_sha256 ((ut8 *)content, r_str_len_utf8 (content) * sizeof (ut8));
 }
 
-static char *read_until(const char *str, const char *find) {
-	char *ret;
-	char *f = strstr (str, find);
-	if (!f) {
-		return NULL;
-	}
-	ret = malloc (f - str + 1);
-	if (!ret) {
-		return NULL;
+static void free_blobs (RList *blobs) {
+	RListIter *iter;
+	RvcBlob *blob;
+	r_list_foreach (blobs, iter, blob) {
+		free (blob->fhash);
+		free (blob->fname);
 	}
-	r_str_ncpy(ret, str, f - str + 1);
-	return ret;
+	r_list_free (blobs);
 }
 
-static char *p2rvcp(const Rvc *repo, const char *path) {
-	char *ret;
-	char *p;
-	char *rp = read_until (repo->path, R_SYS_DIR ".rvc");
-	if (!rp) {
+static char *absp2rp(const char *absp, const char *rp) {
+	char *arp = r_file_abspath (rp);
+	if (!arp) {
 		return NULL;
 	}
-	if (!r_file_is_abspath (path)) {
-		p = r_file_abspath (path);
-	} else {
-		p = r_str_new (path);
-	}
-	if (!p) {
-		free (rp);
+	if (r_str_len_utf8 (arp) < r_str_len_utf8 (rp)) {
+		free (arp);
 		return NULL;
 	}
-	ret = r_str_new (p + r_str_len_utf8 (rp) + 1);
-	free (p);
-	free (rp);
-	return ret;
-}
-
-static RvcBranch *branch_by_name(Rvc *repo, const char *name) {
-	RListIter *iter;
-	RvcBranch *b;
-	r_list_foreach (repo->branches, iter, b) {
-		if  (!r_str_cmp (name, b->name, r_str_len_utf8 (b->name) * sizeof (ut8))) {
-			return b;
-		}
-	}
-
-	return NULL;
+	return r_str_new (absp + r_str_len_utf8 (arp));
 }
 
-static bool write_commit(Rvc *repo, RvcBranch *b, RvcCommit *commit) {
-	char *commit_path, *commit_string;
-	char *prev_path;
-	FILE *prev_file, *commit_file;
-	RListIter *iter;
+static bool bfadd(RList *dst, const char *path, const char *rp) {
 	RvcBlob *blob;
-	commit_string = r_str_newf ("author:%s\nmessage:%s\nntimestamp:%" PRId64"\n----",
-			commit->author, commit->message, commit->timestamp);
-	r_return_val_if_fail (commit_string, false);
-	r_list_foreach (commit->blobs, iter, blob) {
-		char *tmp = r_str_appendf (commit_string, "\nblob:%s:%s",
-				blob->fname, blob->hash);
-		if (!tmp) {
-			return false;
-		}
-		commit_string = tmp;
+	char *absp;
+	char *blob_path;
+	blob = R_NEW (RvcBlob);
```

Comment:
always define and assign in the same line if possible

Reply:
when is is it possible?

Reply:
i mean `RvcBlob *blob = R_NEW0 (RvcBlob);`